From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Chris Fallin <chris@cfallin.org>
Date: Tue, 22 Aug 2023 10:55:18 -0700
Subject: [PATCH] JS shell on WASI: add basic Wizer integration

--enable-js-shell-wizer (JS_SHELL_WIZER): Wizer pre-initialisation of
the standalone JS shell on WASI, js/src/shell/wizer.cpp.

Source commits in YurtOS/firefox (yurt dff5cf534448):
  1e77c0d2599e JS shell on WASI: add basic Wizer integration for standalone testing.
    (all included paths except third_party/wizer)
---
 js/moz.configure       | 16 +++++++++
 js/src/shell/js.cpp    | 81 +++++++++++++++++++++++++++++-------------
 js/src/shell/jsshell.h | 15 ++++++++
 js/src/shell/moz.build |  2 ++
 js/src/shell/wizer.cpp | 59 ++++++++++++++++++++++++++++++
 5 files changed, 148 insertions(+), 25 deletions(-)
 create mode 100644 js/src/shell/wizer.cpp

diff --git a/js/moz.configure b/js/moz.configure
index a363235983ab..1c92f37e74f6 100644
--- a/js/moz.configure
+++ b/js/moz.configure
@@ -57,6 +57,22 @@ def js_disable_shell(value):
 
 set_config("JS_DISABLE_SHELL", js_disable_shell)
 
+option(
+    "--enable-js-shell-wizer",
+    default=False,
+    when=js_standalone,
+    help="{Support|Do not support} wizening of the JS shell",
+)
+
+
+@depends("--enable-js-shell-wizer", when=js_standalone)
+def js_shell_wizer(value):
+    if value:
+        return True
+
+
+set_define("JS_SHELL_WIZER", js_shell_wizer)
+
 set_define("JS_64BIT", depends(target)(lambda t: t.bitness == 64 or None))
 
 set_define("JS_PUNBOX64", depends(target)(lambda t: t.bitness == 64 or None))
diff --git a/js/src/shell/js.cpp b/js/src/shell/js.cpp
index bcaba8febee8..b4193feef5f3 100644
--- a/js/src/shell/js.cpp
+++ b/js/src/shell/js.cpp
@@ -44,6 +44,7 @@
 #endif
 #include <ctime>
 #include <math.h>
+#include <optional>
 #ifndef __wasi__
 #  include <signal.h>
 #endif
@@ -12361,7 +12362,8 @@ ShellContext::~ShellContext() {
   MOZ_ASSERT(offThreadJobs.empty());
 }
 
-static int Shell(JSContext* cx, OptionParser* op) {
+static int Shell(JSContext* cx, OptionParser* op,
+                 MutableHandleObject lastGlobal) {
 #ifdef JS_STRUCTURED_SPEW
   cx->spewer().enableSpewing();
 #endif
@@ -12413,6 +12415,7 @@ static int Shell(JSContext* cx, OptionParser* op) {
     if (!glob) {
       return 1;
     }
+    lastGlobal.set(glob.get());
 
     JSAutoRealm ar(cx, glob);
 
@@ -12660,7 +12663,13 @@ static bool SetGCParameterFromArg(JSContext* cx, char* arg) {
   return true;
 }
 
-int main(int argc, char** argv) {
+// Special-case stdout and stderr. We bump their refcounts to prevent them
+// from getting closed and then having some printf fail somewhere.
+RCFile rcStdout(stdout);
+RCFile rcStderr(stderr);
+
+Variant<JSAndShellContext, int> js::shell::ShellMain(int argc, char** argv,
+                                                     bool retainContext) {
   PreInit();
 
   sArgc = argc;
@@ -12670,13 +12679,8 @@ int main(int argc, char** argv) {
 
   setlocale(LC_ALL, "");
 
-  // Special-case stdout and stderr. We bump their refcounts to prevent them
-  // from getting closed and then having some printf fail somewhere.
-  RCFile rcStdout(stdout);
   rcStdout.acquire();
-  RCFile rcStderr(stderr);
   rcStderr.acquire();
-
   SetOutputFile("JS_STDOUT", &rcStdout, &gOutFile);
   SetOutputFile("JS_STDERR", &rcStderr, &gErrFile);
 
@@ -12688,27 +12692,27 @@ int main(int argc, char** argv) {
 
   OptionParser op("Usage: {progname} [options] [[script] scriptArgs*]");
   if (!InitOptionParser(op)) {
-    return EXIT_FAILURE;
+    return AsVariant(EXIT_FAILURE);
   }
 
   switch (op.parseArgs(argc, argv)) {
     case OptionParser::EarlyExit:
-      return EXIT_SUCCESS;
+      return AsVariant(EXIT_SUCCESS);
     case OptionParser::ParseError:
       op.printHelp(argv[0]);
-      return EXIT_FAILURE;
+      return AsVariant(EXIT_FAILURE);
     case OptionParser::Fail:
-      return EXIT_FAILURE;
+      return AsVariant(EXIT_FAILURE);
     case OptionParser::Okay:
       break;
   }
 
   if (op.getHelpOption()) {
-    return EXIT_SUCCESS;
+    return AsVariant(EXIT_SUCCESS);
   }
 
   if (!SetGlobalOptionsPreJSInit(op)) {
-    return EXIT_FAILURE;
+    return AsVariant(EXIT_FAILURE);
   }
 
   if (!JS::SetLoggingInterface(shellLoggingInterface)) {
@@ -12719,7 +12723,7 @@ int main(int argc, char** argv) {
   // Start the engine.
   if (const char* message = JS_InitWithFailureDiagnostic()) {
     fprintf(gErrFile->fp, "JS_Init failed: %s\n", message);
-    return 1;
+    return AsVariant(1);
   }
 
   // `selfHostedXDRBuffer` contains XDR buffer of the self-hosted JS.
@@ -12731,11 +12735,11 @@ int main(int argc, char** argv) {
   auto shutdownEngine = MakeScopeExit([] { JS_ShutDown(); });
 
   if (!SetGlobalOptionsPostJSInit(op)) {
-    return EXIT_FAILURE;
+    return AsVariant(EXIT_FAILURE);
   }
 
   if (!InitSharedObjectMailbox()) {
-    return EXIT_FAILURE;
+    return AsVariant(EXIT_FAILURE);
   }
 
   JS::SetProcessBuildIdOp(ShellBuildId);
@@ -12743,7 +12747,7 @@ int main(int argc, char** argv) {
   /* Use the same parameters as the browser in xpcjsruntime.cpp. */
   JSContext* const cx = JS_NewContext(JS::DefaultHeapMaxBytes);
   if (!cx) {
-    return 1;
+    return AsVariant(1);
   }
 
   // Register telemetry callbacks.
@@ -12755,11 +12759,11 @@ int main(int argc, char** argv) {
   UniquePtr<ShellContext> sc =
       MakeUnique<ShellContext>(cx, ShellContext::MainThread);
   if (!sc || !sc->registerWithCx(cx)) {
-    return 1;
+    return AsVariant(1);
   }
 
   if (!SetContextOptions(cx, op)) {
-    return 1;
+    return AsVariant(1);
   }
 
   JS_SetTrustedPrincipals(cx, &ShellPrincipals::fullyTrusted);
@@ -12772,7 +12776,7 @@ int main(int argc, char** argv) {
   bufferStreamState = js_new<ExclusiveWaitableData<BufferStreamState>>(
       mutexid::BufferStreamState);
   if (!bufferStreamState) {
-    return 1;
+    return AsVariant(1);
   }
   auto shutdownBufferStreams = MakeScopeExit([] {
     ShutdownBufferStreams();
@@ -12814,7 +12818,7 @@ int main(int argc, char** argv) {
   }
 
   if (!JS::InitSelfHostedCode(cx, xdrSpan, xdrWriter)) {
-    return 1;
+    return AsVariant(1);
   }
 
   EnvironmentPreparer environmentPreparer(cx);
@@ -12828,13 +12832,14 @@ int main(int argc, char** argv) {
     if (!WasmCompileAndSerialize(cx)) {
       // Errors have been printed directly to stderr.
       MOZ_ASSERT(!cx->isExceptionPending());
-      return EXIT_FAILURE;
+      return AsVariant(EXIT_FAILURE);
     }
 #endif
-    return EXIT_SUCCESS;
+    return AsVariant(EXIT_SUCCESS);
   }
 
-  result = Shell(cx, &op);
+  RootedObject lastGlobal(cx);
+  result = Shell(cx, &op, &lastGlobal);
 
 #ifdef DEBUG
   if (OOM_printAllocationCount) {
@@ -12842,9 +12847,29 @@ int main(int argc, char** argv) {
   }
 #endif
 
-  return result;
+  if (retainContext) {
+    shutdownEngine.release();
+    destroyCx.release();
+    shutdownBufferStreams.release();
+    shutdownShellThreads.release();
+
+    JSAndShellContext ret{cx, lastGlobal.get(), std::move(sc),
+                          std::move(selfHostedXDRBuffer)};
+    return AsVariant(std::move(ret));
+  } else {
+    return AsVariant(result);
+  }
 }
 
+// N.B.: When Wizer support is enabled, a separate main() is used.
+#ifndef JS_SHELL_WIZER
+
+int main(int argc, char** argv) {
+  return ShellMain(argc, argv, /* returnContext = */ false).as<int>();
+}
+
+#endif  // !JS_SHELL_WIZER
+
 bool InitOptionParser(OptionParser& op) {
   op.setDescription(
       "The SpiderMonkey shell provides a command line interface to the "
@@ -13620,6 +13645,12 @@ bool SetGlobalOptionsPostJSInit(const OptionParser& op) {
         JS::DelazificationOption::ParseEverythingEagerly;
   }
 
+  // Likewise, if we have Wizer support built into the shell, we
+  // unconditionally parse everything eagerly.
+#ifdef JS_SHELL_WIZER
+  defaultDelazificationMode = JS::DelazificationOption::ParseEverythingEagerly;
+#endif
+
   if (const char* xdr = op.getStringOption("selfhosted-xdr-path")) {
     shell::selfHostedXDRPath = xdr;
   }
diff --git a/js/src/shell/jsshell.h b/js/src/shell/jsshell.h
index 85174e25a9d1..35e04fae1a8d 100644
--- a/js/src/shell/jsshell.h
+++ b/js/src/shell/jsshell.h
@@ -19,6 +19,7 @@
 #include "threading/LockGuard.h"
 #include "threading/Mutex.h"
 #include "threading/Thread.h"
+#include "util/CompleteFile.h"  // js::FileContents
 #include "vm/GeckoProfiler.h"
 #include "vm/Monitor.h"
 
@@ -259,11 +260,25 @@ struct ShellContext {
   JS::PersistentRooted<ObjectVector> taskCallbacks;
 };
 
+// A JSContext, the associated global object, the ShellContext, and
+// other miscellaneous bits that need to live as long as the
+// shell. This is used to wrap up the shell state when it needs to be
+// persisted (e.g., in a Wizer snapshot).
+struct JSAndShellContext {
+  JSContext* cx;
+  JSObject* glob;
+  UniquePtr<ShellContext> shellCx;
+  mozilla::Maybe<js::FileContents> selfHostedXDRBuffer;
+};
+
 extern ShellContext* GetShellContext(JSContext* cx);
 
 [[nodiscard]] extern bool PrintStackTrace(JSContext* cx,
                                           JS::Handle<JSObject*> stackObj);
 
+mozilla::Variant<JSAndShellContext, int> ShellMain(int argc, char** argv,
+                                                   bool retainContext);
+
 } /* namespace shell */
 } /* namespace js */
 
diff --git a/js/src/shell/moz.build b/js/src/shell/moz.build
index d7fa408fc38f..e8ca9e67a1e2 100644
--- a/js/src/shell/moz.build
+++ b/js/src/shell/moz.build
@@ -22,6 +22,7 @@ UNIFIED_SOURCES += [
     "OSObject.cpp",
     "ShellModuleObjectWrapper.cpp",
     "WasmTesting.cpp",
+    "wizer.cpp",
 ]
 
 if CONFIG["FUZZING_INTERFACES"]:
@@ -39,6 +40,7 @@ DEFINES["EXPORT_JS_API"] = True
 LOCAL_INCLUDES += [
     "!..",
     "..",
+    "/third_party/wizer",
 ]
 
 OS_LIBS += CONFIG["EDITLINE_LIBS"]
diff --git a/js/src/shell/wizer.cpp b/js/src/shell/wizer.cpp
new file mode 100644
index 000000000000..3e0a3fa91a11
--- /dev/null
+++ b/js/src/shell/wizer.cpp
@@ -0,0 +1,59 @@
+/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 2 -*-
+ * vim: set ts=8 sts=2 et sw=2 tw=80:
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/* Support for Wizer-based snapshotting of the JS shell, when built
+ * for a Wasm target (i.e., running inside a Wasm module). */
+
+#include "js/CallAndConstruct.h"  // JS_CallFunctionName
+#include "shell/jsshell.h"
+
+using namespace js;
+using namespace js::shell;
+
+#ifdef JS_SHELL_WIZER
+
+#  include <wizer.h>
+
+static std::optional<JSAndShellContext> wizenedContext;
+
+static void WizerInit() {
+  const int argc = 1;
+  char* argv[2] = {strdup("js"), NULL};
+
+  auto ret = ShellMain(argc, argv, /* retainContext = */ true);
+  if (!ret.is<JSAndShellContext>()) {
+    fprintf(stderr, "Could not execute shell main during Wizening!\n");
+    abort();
+  }
+
+  wizenedContext = std::move(ret.as<JSAndShellContext>());
+}
+
+WIZER_INIT(WizerInit);
+
+int main(int argc, char** argv) {
+  (void)argc;
+  (void)argv;
+
+  if (wizenedContext) {
+    JSContext* cx = wizenedContext.value().cx;
+    RootedObject glob(cx, wizenedContext.value().glob);
+
+    JSAutoRealm ar(cx, glob);
+
+    // Look up a function called "main" in the global.
+    JS::Rooted<JS::Value> ret(cx);
+    if (!JS_CallFunctionName(cx, cx->global(), "main",
+                             JS::HandleValueArray::empty(), &ret)) {
+      fprintf(stderr, "Failed to call main() in Wizened JS source!\n");
+      abort();
+    }
+  } else {
+    return ShellMain(argc, argv, /* returnContext = */ false).as<int>();
+  }
+}
+
+#endif  // JS_SHELL_WIZER
